### PR TITLE
Review for 'tree refactor': filtering, ordering and incremental search

### PR DESCRIPTION
**Pull request by shaneporter**
#2448 - tree filtering based on time
#2391 - incremental search
#2387 - time sorted notebook tree

I apologise that this is such a huge PR (and that many of the earlier commits are described simply as 'refactor'); I am aware of some issues, but also suspect that some other areas may not be bulletproof:

- Incremental search's SOLR query needs tweaking, but hopefully what's returned _is_ displayed correctly, and users can open the notebooks
- Imported notebooks may not display if they don't match the current filter

### CODE REVIEW

File: htdocs/js/tree/notebook_tree_model.js [hunk 1]
```diff
@@ -0,0 +1,1691 @@
...
+
+            // initial assignment: 
+            this.tree_data_ = data;
+
+            this.update_filter({
+                prop: 'tree_filter_date',
+                value: opts['tree_filter_date']
+            });
+
+            this.update_sort_type(opts['tree_sort_order'], true);
```
Comment (gordonwoodhull) [inline htdocs/js/tree/notebook_tree_model.js]: remember that these `opts` will be undefined the first time the code is run - probably worth testing with all options set to `null`...

**Comment by gordonwoodhull** [comment]
As can be expected, I got pulled into some other "urgent" work and wasn't able to review this week.

I think I will just merge once it works well enough, since this is too huge to wait much longer. I trust that you know what you're doing, and it will just take time to absorb all these changes.

Once I got around the error above by manipulating data in the debugger, sort and filter worked great in a quick test! Excellent work, @shaneporter.

The layout could use tightening up. This may be a Chrome/MacOS issue but there is some line wrapping which makes it tall.

image

I'd expect it also to stick to the top and not scroll, but I'm open to opinions on that. I think if it is just one line tall as intended, it won't take up too much space. When digging in All Notebooks it's quite common to be scrolled away, and we wouldn't want to forget what the sort/filter are.

I'm not sure if the white border helps - guess this will be clear when it's the right height.

It seems strange to me to pop up a dialog for incremental search? Shouldn't that just be another line inside the notebook tree? And of course I'm missing some search APIs so I can't test it.

Merge next week is going to be a lot of fun!

**Comment by gordonwoodhull** [comment]
Ah, I guess the dialog is because we are not incrementally searching within the tree UI. Hmm. I'm not a huge fan of modal dialogs, but I'm not sure what to do. May be too late to do anything about this now.

Also, <kbd>cmd p</kbd> is the print hotkey, which we should not steal. Can we use <kbd>alt s</kbd>?

**Comment by gordonwoodhull** [comment]
This works almost perfectly, and we're running out of time, so I'll go ahead and merge today.

I'll let you know separately about any issues.

I like the new design of the incremental search dialog, too - attaching it to the top makes sense.

Thanks @shaneporter!

**Comment by gordonwoodhull** [comment]
@shaneporter, @dougmet, I'm slightly confused about merging this and #2459, since they both modified search.R but I know you guys were working together on this.

What I will try to do (and I know it's too late in the day/week to get any feedback until Monday, so I'm partly posting this so that I remember what I did):

1.  Keep all of Doug's changes, which remove search.R and put that stuff into rcloud.solr
2. Also keep `rcloud.search.description` (restoring search.R) but change it so that it references the rcloud.solr package instead of directly calling `.solr.get`

It looks like `rcloud.search.description` is a bit higher-level than the rest of rcloud.solr, so it should stay in rcloud.support, but any opinions are welcome here.

@s-u and I are in the same place today so I'm hoping to get our dev server up and running.

Thanks!

**Comment by gordonwoodhull** [comment]
Ah, okay, I was just confused in the middle of the merge, everything fits together now!

**Comment by shaneporter** [comment]
@dougmet  wrote the `rcloud.search.description` method for the incremental search so I had something to call, but I anticipated there being some changes from 'my' side once the merge had been done. Thanks.

**Comment by gordonwoodhull** [comment]
Thanks, I think I understand. It is moved into rcloud.solr. 

I just wasn't finished with the merge when I thought I was!

**Pull request merged**